Show battery level for selected devices in Power Monitoring Console

ArtisticRoomba opened the pull request:
## About the PR
Adds a charge percentage and charge bar to applicable devices in the Power Monitoring Console.

This is a revival of derelict PR #26793, originally coded by Crotalus. I've tried to make it look decent, this is my first PR touching any drop of C#.

## Why / Balance
Extremely needed QOL feature. This tells engineers battery levels across the station and can prove EXTREMELY USEFUL to know. Really gives you a good idea as to where power is moving around in the station, and how the grid behaves as a whole.

## Technical details
* Added battery level indicators to the `PowerMonitoringButton` class, including a progress bar and percentage label for visual representation of battery levels.

* Refactored `GetPrimaryPowerValues` into `PowerStats` class to encapsulate power statistics.

* Updated the `PowerMonitoringConsoleEntry` struct to include a nullable `BatteryLevel` property to store battery level information. (`Content.Shared/Power/SharedPowerMonitoringConsoleComponent.cs`)

* Adds the `BatteryLevel` property to the `PowerMonitoringConsoleEntry` objects that are being created and modified within the `GetSourcesForNode` and `GetLoadsForNode` methods.

## Media
![Content Client_weC4i7G7fx](https://github.com/user-attachments/assets/3c9ef2cb-a88d-4a8b-84de-052b33be5b3c)

![Content Client_ngiUnjU4RL](https://github.com/user-attachments/assets/32f0e3e7-2236-429a-aa7d-63bf438066f0)

![Content Client_nKd9VEQ4FD](https://github.com/user-attachments/assets/1a7bbd5a-f681-4cd5-9335-8918aff95388)

![Content Client_2R36TbOB7F](https://github.com/user-attachments/assets/c365f374-1de4-4d8e-bfcc-61553f6d68dd)

![image](https://github.com/user-attachments/assets/dcc5ce4d-2c29-406a-91d1-42b82160c0b1)

## Requirements
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.


## Breaking changes
Shouldn't be any.

**Changelog**
:cl: Crotalus, ArtisticRoomba
- add: Battery levels of applicable devices are now shown in the Power Monitoring Console!

PJB3005 commented on Resources/Locale/en-US/components/power-monitoring-component.ftl:
```diff
@@ -13,6 +13,7 @@ power-monitoring-window-unknown-location = Unknown location
 power-monitoring-window-total-sources = Total generator output
 power-monitoring-window-total-battery-usage = Total battery usage
 power-monitoring-window-total-loads = Total network loads
+power-monitoring-window-battery-level = {$percentage}%
```
This isn't used.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -104,6 +105,28 @@ public void UpdateWindowEntryButton(NetEntity netEntity, PowerMonitoringButton b
+            button.BatteryPercentage.Visible = true;
+
+            // Set progress bar color based on percentage
+            var color = Color.FromHsv(new Vector4((batteryPercentage / 100f) * 0.33f, 1, 1, 1));
```
Why not use `entry.BatteryLevel.Value` directly? You're dividing and then multiplying by 100 here.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -104,6 +105,28 @@ public void UpdateWindowEntryButton(NetEntity netEntity, PowerMonitoringButton b
+            button.BatteryLevel.Value = batteryPercentage;
+            button.BatteryLevel.Visible = true;
+
+            button.BatteryPercentage.Text = $"{batteryPercentage}%";
```
You should use `.ToString("P0")` instead so this localizes properly.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -478,6 +505,31 @@ public PowerMonitoringButton()
+            HorizontalAlignment = HAlignment.Right,
+            SetWidth = 60f,
+            Margin = new Thickness(15, 7, 0, 7),
+            MinValue = 0,
+            MaxValue = 100,
+            Page = 0,
+            Value = 0,
```
You're just setting all of these to the default, there's no point to.

Also set `MaxValue` to 1 so you can avoid multiplying by 100 when setting the value.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -478,6 +505,31 @@ public PowerMonitoringButton()
+        BatteryPercentage = new Label()
+        {
+            HorizontalAlignment = HAlignment.Right,
+            SetWidth = 45f,
```
You should probably set a MinWidth instead so the layout doesn't break when we get the ability for users to select custom fonts.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -478,6 +505,31 @@ public PowerMonitoringButton()
 
+        BatteryLevel = new ProgressBar()
+        {
+            HorizontalAlignment = HAlignment.Right,
```
This doesn't do anything.

PJB3005 commented on Content.Client/Power/PowerMonitoringWindow.xaml.Widgets.cs:
```diff
@@ -478,6 +505,31 @@ public PowerMonitoringButton()
+
+        BatteryPercentage = new Label()
+        {
+            HorizontalAlignment = HAlignment.Right,
```
This doesn't do anything.

PJB3005 commented on Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs:
```diff
@@ -990,4 +1005,13 @@ private void RefreshPowerMonitoringCableNetworks(EntityUid uid, PowerMonitoringC
         Dirty(uid, component);
     }
+
+    private sealed class PowerStats
```
This should be a struct.

PJB3005 reviewed (changes requested):
> I did some debugging and saw that the PowerMonitoringConsoleEntry[] struct list never has a BatteryLevel float (null), while PowerMonitoringConsoleEntry does when applicable. Because the value is always null, the battery info is never drawn.
> 
> I think this is boiled down to it not being recorded somewhere in PowerMonitoringConsoleSystem, however I don't have experience nor the time right now to go further than what I've already done.

There are other places in the system code where these `PowerMonitoringConsoleEntry` values are being made, you should update them too.

ArtisticRoomba commented:
> There are other places in the system code where these `PowerMonitoringConsoleEntry` values are being made, you should update them too.

I addressed your review, however I admittedly don't know where exactly these values are being made, after I spent some time searching in the relevant system. I'll look around some more. Sorry.

ArtisticRoomba commented:
NEVERMIND
![image](https://github.com/user-attachments/assets/423c928f-c2c4-4cdb-bc74-121d896fb1fd)

slarticodefast reviewed (approved):
Blessed PR.
Tested and works as intended.
